src/clusterfuzz/_internal/bot/tasks/utasks/__init__.py

```diff
@@ -53,17 +55,70 @@ def _timestamp_now() -> Timestamp:
   return ts
 
 
-def _record_e2e_duration(start: Timestamp, utask_module, job_type: str,
-                         subtask: _Subtask, mode: _Mode):
-  duration_secs = (time.time_ns() - start.ToNanoseconds()) / 10**9
-  monitoring_metrics.UTASK_E2E_DURATION_SECS.add(
-      duration_secs, {
-          'task': task_utils.get_command_from_module(utask_module.__name__),
-          'job': job_type,
-          'subtask': subtask.value,
-          'mode': mode.value,
-          'platform': environment.platform(),
-      })
+class _MetricRecorder(contextlib.AbstractContextManager):
+  """Records task execution metrics, even in case of error and exceptions.
+
+  Members:
+    start_time_ns (int): The time at which this recorder was constructed, in
+      nanoseconds since the Unix epoch.
+  """
+
+  def __init__(self, subtask: _Subtask, mode: _Mode):
+    self.start_time_ns = time.time_ns()
+    self._subtask = subtask
+    self._mode = mode
+    self._labels = None
+
+    if subtask == _Subtask.PREPROCESS:
+      self._preprocess_start_time_ns = self.start_time_ns
```

Comment:
Why use nanoseconds instead of seconds?

Reply:
I think I was being pedantic, and this will not matter at all. I can change it back if you prefer.